src/Microsoft.AspNetCore.Mvc.TagHelpers/CacheTagHelper.cs

```diff
@@ -156,87 +56,71 @@ public override int Order
             {
                 throw new ArgumentNullException(nameof(output));
             }
+            
+            IHtmlContent content = null;
 
-            IHtmlContent result = null;
             if (Enabled)
             {
                 var key = GenerateKey(context);
-                if (!MemoryCache.TryGetValue(key, out result))
+                MemoryCacheEntryOptions options;
+    
+                while(content == null)
                 {
-                    // Create an entry link scope and flow it so that any tokens related to the cache entries
-                    // created within this scope get copied to this scope.
-                    using (var link = MemoryCache.CreateLinkingScope())
+                    Task<IHtmlContent> result = null;
+                    
+                    if (!MemoryCache.TryGetValue(key, out result))
                     {
-                        var content = await output.GetChildContentAsync();
-
-                        var stringBuilder = new StringBuilder();
-                        using (var writer = new StringWriter(stringBuilder))
+                        // Create an entry link scope and flow it so that any tokens related to the cache entries
+                        // created within this scope get copied to this scope.
+                        options = GetMemoryCacheEntryOptions();
+
+                        var tcs = new TaskCompletionSource<IHtmlContent>();
+                         
+                        MemoryCache.Set(key, tcs.Task, options);
+                        
+                        using (var link = MemoryCache.CreateLinkingScope())
                         {
-                            content.WriteTo(writer, HtmlEncoder);
+                            result = ProcessContentAsync(output);
+                            content = await result;
+                            options.AddEntryLink(link);
                         }
-
-                        result = new StringBuilderHtmlContent(stringBuilder);
-                        MemoryCache.Set(key, result, GetMemoryCacheEntryOptions(link));
+                        
+                        MemoryCache.Set(key, result, options);
+                        tcs.SetResult(null);
                     }
+                    else
+                    {
+                        content = await result;    
+                    }                    
                 }
             }
 
             // Clear the contents of the "cache" element since we don't want to render it.
             output.SuppressOutput();
-            if (Enabled)
-            {
-                output.Content.SetContent(result);
-            }
-            else
+
+            if (!Enabled)
```

Comment:
Are you sure the `SuppressOutput` call won't revert it? I kept the same flow as the previous version.